### PR TITLE
Support ppx using the typer such as ppx_import

### PR DESCRIPTION
**Pull request by ghost**
## Overview

This PR allows to use ppx rewriters such as ppx_import that are using the OCaml typer. To do so it adds a new preprocessing specification `(pps_using_typer ...)`. It is similar to `(pps ...)` except that instead of preprocessing files only once, dune will use the `-ppx` option of ocamldep, ocamlc and ocamlopt.

## Implementation

I added a new field `pp : (unit, string list) Build.t` to `Module.t`. The `Preprocessing` module fills this field with an arrow that evaluates to `["-ppx"; "<ppx command>"]` and it is then used by `Ocamldep` and `Module_compilation`.

## Rationale for having both pps and pps_using_typer

To be sure that it is worth having two code paths, I tried replacing `pps` by `pps_using_typer` in core and compared the compilation times with `-j1`. The slow-down is indeed noticeable: about 10%, so systematically using `-ppx` is not acceptable.

## Possible improvement

The fact that a ppx rewriter is using the typer is really a property of the rewriter itself, so ideally we should mark such ppx rewriters with `(kind (ppx_rewriter (using_typer)))`. I'm not planning to do the necessary work to support this and leave that part to someone else.

### CODE REVIEW

**Comment by diml** [comment]
This fixes #193, /cc @xguerin

**Comment by emillon** [comment]
Before diving into the implementation, I have a question about the name and interface: is it really about the ppx using the typechecker, or just a way to use `-ppx` rather than a driver? If there are other reasons to use `-ppx`, it will be confusing to use `using_typer` to use that code path.

**Comment by diml** [comment]
It is still using the same driver. The difference is simply:

- build process with pps:
  + generate a custom `ppx.exe`
  + `ppx.exe --dump-ast file.ml -o file.pp.ml`
  + `ocamldep file.pp.ml > file.ml.d`
  + `ocamlc -c file.pp.ml`
  + `ocamlopt -c file.pp.ml`
- build process with pps_using_typer:
  + generate a custom `ppx.exe`
  + `ocamldep -pp "ppx.exe --as-ppx" file.ml > file.ml.d`
  + `ocamlc -pp "ppx.exe --as-ppx" -c file.ml`
  + `ocamlopt -pp "ppx.exe --as-ppx" -c file.ml`

The first process is faster, so the only reason I can think of one would want to use the second one is when using a ppx that is using the OCaml typer.

**Comment by emillon** [comment]
Thanks, that explanation is helpful. Do you think it would be worth adding it to the docs?

I wonder if we can describe it in terms of passes: the default strategy works in two passes: preprocess, then compile, but this kind of ppx requires doing it all at once in a single pass. So, would a name like `(pps_in_single_pass ...)` work?

**Comment by diml** [comment]
Adding something to the doc seems good. Regarding the name, if we rename `pps_using_typer` to `pps_in_single_pass`, it makes me thing that `pps` means `pps_in_multiple_pass` and then is slower, which is not the case. Maybe we could rename `pps_using_typer` to `staged_pps` and describe things like this:

## Doc

Dune normally assumes that the compilation pipeline is sequenced as follow:
- code generation (including preprocessing)
- dependency analisys
- compilation

Dune uses this fact to optimize the pipeline and in particular share the result of code generation and preprocessing between the dependency analisys and compilation phases. However, some specific code generators or preprocessors require feedback from the compilation phase. As a result they must be applied in stages as follow:
- first stage of code geneneration
- dependency analisys
- second step of code generation in parallel with compilation

This is the case for ppx rewriters using the OCaml typer for instance. When using such ppx rewriters, you must use `staged_pps` instead of `pps` in order to force Dune to use the second pipeline, which is slower but necessary in this case.

**Comment by emillon** [comment]
That seems good to me. I don't have a strong opinion about the name, but `staged_pps` seems a bit more neutral so I think I prefer this.

**Review by rgrinberg — APPROVED** [review]
Looks good. I have no preference to the names either.

**Comment by diml** [comment]
I made the change

File: doc/dune-files.rst [hunk 1]
```diff
@@ -1107,9 +1107,31 @@ Dune accepts three kinds of preprocessing:
...
+Dune normally assumes that the compilation pipeline is sequenced as
+follow:
+
+- code generation (including preprocessing)
+- dependency analisys
+- compilation
+
+Dune uses this fact to optimize the pipeline and in particular share
+the result of code generation and preprocessing between the dependency
+analisys and compilation phases. However, some specific code
```
Comment (emillon) [inline doc/dune-files.rst]: mini typo here: analysis

File: doc/dune-files.rst [hunk 2]
```diff
@@ -1107,9 +1107,31 @@ Dune accepts three kinds of preprocessing:
...
+generators or preprocessors require feedback from the compilation
+phase. As a result they must be applied in stages as follow:
```
Comment (emillon) [inline doc/dune-files.rst]: "as follows"

File: doc/dune-files.rst [hunk 3]
```diff
@@ -1107,9 +1107,31 @@ Dune accepts three kinds of preprocessing:
...
+
+- first stage of code geneneration
+- dependency analisys
```
Comment (emillon) [inline doc/dune-files.rst]: analysis

**Review by emillon — APPROVED** [review]
There are some minor typos but otherwise it's 👍

**Comment by diml** [comment]
Thanks, I fixed the typos

**Pull request merged**